### ISSUE TITLE
100% CPU after completing a recurring job

### ISSUE THREAD

**Issue by onetsmr** [HangfireIO/Hangfire#1972]
Hi, 

We use Hangfire 1.7.27 + Hagfire.Pro 2.3.0.0 + Hagfire.Pro.Redis 2.8.12.0

We are facing 100% CPU usage while hangfire is not processing any jobs after completing a recurring job
![01](https://user-images.githubusercontent.com/7472809/142371325-32b90b50-15ef-4afa-98cb-4e7a94b9b3bd.png)

How does this happen:
1. If restart the IIS then CPU load is 1-2%
2. If the only recurring job will be executed at least once then CPU load is 98-99% and does not decrease for a long time

I've collected trace at this moment on the server using dotTrace
![02](https://user-images.githubusercontent.com/7472809/142371614-e568103d-dae4-433e-add0-6d8b95e5a8b5.png)
RunWorkLoop in DedicatedThreadPoolPipeScheduler takes 50% time
Wait in CancellationTokenExtentions takes 20% time

This is RunWorkLoop code
https://github.com/mgravell/Pipelines.Sockets.Unofficial/blob/19b141f30fe2485dd79993be0ac87a9fc6b34b74/src/Pipelines.Sockets.Unofficial/DedicatedThreadPoolPipeScheduler.cs#L166

**Comment by odinserj** [comment]
Thanks for the screenshots and profiling. Please note that Hangfire.Pro.Redis doesn't use Pipelines.Sockets.Unofficial, because based on the internalised [custom branch](https://github.com/odinserj/StackExchange.Redis) of StackExchange.Redis 1.2.6 (so it doesn't matter what version is specified in the project itself). Also, `CancellationTokenExtensions.Wait` method is based on `WaitHandle.WaitOne` method that doesn't consume the CPU itself (despite shown in the profiler).

I'm more interested in those Thread Pool threads as they show a lot of activities. Is it still possible to examine their top methods and send some screenshots?

**Comment by odinserj** [comment]
Also could you please show inner details for the following call?

![image](https://user-images.githubusercontent.com/1078718/142440448-c5758041-b12a-4fc7-9017-5fc14df267d8.png)

**Comment by onetsmr** [comment]
We download and use Hagfire.Pro.Redis, not Hangfire.Pro.Redis.SEv2. I will try SEv2.

Thread Pools has 77% native code
![Screenshot_2](https://user-images.githubusercontent.com/7472809/142448368-bec3f33b-6a83-4b1c-9f00-ad7fddb3be10.png)
![Screenshot_3](https://user-images.githubusercontent.com/7472809/142448384-6a1d6209-a306-4653-92c3-f6015cce6fb1.png)

RunWorkLoop also has calling ntdll
![Screenshot_1](https://user-images.githubusercontent.com/7472809/142448520-e182c6af-8de2-4481-aa10-c5a659c673df.png)

This is IIS 10.0 running on 64bit Windows Server 2016. Web service .net framework 4.6

**Comment by odinserj** [comment]
No, please don't use SEv2 – it's based on StackExchange.Redis 2.X that may be the reason of the issue with high CPU. Hangfire.Pro.Redis is based on the previous version of Redis client that doesn't use Sockets.Pipelines.Unifficial so at least we can isolate things.

**Comment by odinserj** [comment]
Could you tell me what does your recurring job do – I assume it talks with Redis, right?

**Comment by onetsmr** [comment]
Recurring job runs a few child jobs (10-100) based on tenants from the sql database.
Each child job goes to the sql database, takes 10 records, processes them and re run itself while database has rows for processing.
Both types of jobs uses RedLockNet for business locks. And hangfire and RedLockNet goes to the same redis cluster (3 nodes), but uses different keys.

As I understand, I should try to remove things one by one and watch how the situation has changed?

**Comment by onetsmr** [comment]
Clean recurring job - cpu 1-2%
Clean recurring job with redlock - cpu 1-2%

Looks like the reason is not in Hangfire. Sorry.

**Comment by odinserj** [comment]
What if you add some delay after acquiring and before releasing it? Please note it's better to use async Task.Delay or sync Thread.Sleep, depending on the code that's being replaced.

**Comment by odinserj** [comment]
Will appreciate if you can share details for those native calls from profiler (that unfolded fat arrow in the bottom tree on the screenshot related to native calls in thread pool threads looks very interesting). Let's investigate this issue, it looks really strange, since all you have is SQL calls – by the way do you mean SQL Server or MySQL?

**Comment by onetsmr** [comment]
>What if you add some delay after acquiring and before releasing it?
I've tried use BackgroundJob.Schedule(..., TimeSpan.FromSecond(1)) - didn't help

MySQL, okay, I will investigate something myself and return back with questions or answers.

Thank you.

**Comment by onetsmr** [comment]
The problem was in our code - there is a trotting module with semaphores and they are not disposed correctly.
